pkg/controller/nodelink/nodelink_controller.go

```diff
@@ -0,0 +1,488 @@
+	return reconcile.Result{}, nil
+}
+
+// nodeRequestFromMachine returns a reconcile.request for the node backed by the received machine
+func (r *ReconcileNodeLink) nodeRequestFromMachine(o handler.MapObject) []reconcile.Request {
+	klog.V(3).Infof("Watched machine event, finding node to reconcile.Request")
+	// get machine
+	machine := &mapiv1beta1.Machine{}
+	if err := r.client.Get(
+		context.Background(),
+		client.ObjectKey{
+			Namespace: o.Meta.GetNamespace(),
+			Name:      o.Meta.GetName(),
+		},
+		machine,
+	); err != nil {
+		klog.Errorf("No-op: Unable to retrieve machine %s/%s from store: %v", o.Meta.GetNamespace(), o.Meta.GetName(), err)
+		return []reconcile.Request{}
+	}
+
+	if machine.DeletionTimestamp != nil {
+		klog.V(3).Infof("No-op: Machine %q has a deletion timestamp", o.Meta.GetName())
+		return []reconcile.Request{}
+	}
+
+	// find node
+	node, err := r.findNodeFromMachine(machine)
+	if err != nil {
+		klog.Errorf("No-op: Failed to find node for machine %q: %v", machine.GetName(), err)
+		return []reconcile.Request{}
+	}
+	if node != nil {
+		return []reconcile.Request{
+			{
+				NamespacedName: client.ObjectKey{
+					Namespace: node.GetNamespace(),
+					Name:      node.GetName(),
+				},
+			},
+		}
+	}
+
+	klog.V(3).Infof("No-op: Node for machine %q not found", machine.GetName())
+	return []reconcile.Request{}
+}
+
+// findNodeFromMachine tries to a node from by providerIP and fallback to find by IP
+func (r *ReconcileNodeLink) findNodeFromMachine(machine *mapiv1beta1.Machine) (*corev1.Node, error) {
+	if machine == nil {
+		klog.V(3).Infof("Finding node from machine nil, returning...")
+		return nil, nil
+	}
+
+	klog.V(3).Infof("Finding node from machine %q", machine.GetName())
+	node, err := r.findNodeFromMachineByProviderID(machine)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find node from machine %q by ProviderID: %v", machine.GetName(), err)
+	}
+	if node != nil {
+		return node, nil
+	}
+
+	node, err = r.findNodeFromMachineByIP(machine)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find node from machine %q by internal IP: %v", machine.GetName(), err)
+	}
+	return node, nil
+}
+
+func (r *ReconcileNodeLink) findNodeFromMachineByProviderID(machine *mapiv1beta1.Machine) (*corev1.Node, error) {
+	if machine == nil {
+		klog.V(3).Infof("Finding node from machine nil, returning...")
+		return nil, nil
+	}
+
+	klog.V(3).Infof("Finding node from machine %q by providerID", machine.GetName())
+	if machine.Spec.ProviderID == nil {
+		klog.Warningf("Machine %q has no providerID", machine.GetName())
+		return nil, nil
+	}
+
+	nodes, err := r.listNodesByFieldFunc(nodeProviderIDIndex, *machine.Spec.ProviderID)
+	if err != nil {
+		return nil, fmt.Errorf("failed getting node list: %v", err)
+	}
+
+	if len(nodes) > 0 {
+		klog.V(3).Infof("Found node %q for machine %q with providerID %q", nodes[0].GetName(), machine.GetName(), nodes[0].Spec.ProviderID)
+		return nodes[0].DeepCopy(), nil
```

Comment:
should we return the most recently created node?

Reply:
Let's error if len>1, that's an anomaly and error should eventually reconcile towards success